In [1]:
import pandas as pd 

In [2]:
df = pd.read_csv("general_data.csv")
df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [3]:
df = df.drop_duplicates()

In [4]:
df = df.dropna()

In [5]:
df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [6]:
df["Attrition"] = df["Attrition"].replace('Yes', 1)
df["Attrition"] = df["Attrition"].replace('No', 0)

In [7]:
df.head(10)

Age  Attrition     BusinessTravel              Department  \
0   51          0      Travel_Rarely                   Sales   
1   31          1  Travel_Frequently  Research & Development   
2   32          0  Travel_Frequently  Research & Development   
3   38          0         Non-Travel  Research & Development   
4   32          0      Travel_Rarely  Research & Development   
5   46          0      Travel_Rarely  Research & Development   
6   28          1      Travel_Rarely  Research & Development   
7   29          0      Travel_Rarely  Research & Development   
8   31          0      Travel_Rarely  Research & Development   
9   25          0         Non-Travel  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                 6          2  Life Sciences              1           1   
1                10          1  Life Sciences              1           2   
2                17          4          Other              1           3   
3                 2          5  Life Sciences              1           4   
4                10          1        Medical              1           5   
5                 8          3  Life Sciences              1           6   
6                11          2        Medical              1           7   
7                18          3  Life Sciences              1           8   
8                 1          3  Life Sciences              1           9   
9                 7          4        Medical              1          10   

   Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0  Female  ...                 1.0      Y                11              8   
1  Female  ...                 0.0      Y                23              8   
2    Male  ...                 1.0      Y                15              8   
3    Male  ...                 3.0      Y                11              8   
4    Male  ...                 4.0      Y                12              8   
5  Female  ...                 3.0      Y                13              8   
6    Male  ...                 2.0      Y                20              8   
7    Male  ...                 2.0      Y                22              8   
8    Male  ...                 0.0      Y                21              8   
9  Female  ...                 1.0      Y                13              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   
5                 0              28.0                      5               7   
6                 1               5.0                      2               0   
7                 3              10.0                      2               0   
8                 0              10.0                      2               9   
9                 1               6.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  
5                        7                     7  
6                        0                     0  
7                        0                     0  
8                        7                     8  
9                        1                     5  

[10 rows x 24 columns]

In [8]:
attrition_corr = df.corr()[:]['Attrition']

In [9]:
attrition_corr # Correlation Attrition vs Other Variables

Age                       -0.158399
Attrition                  1.000000
DistanceFromHome          -0.009449
Education                 -0.017106
EmployeeCount                   NaN
EmployeeID                -0.004621
JobLevel                  -0.012382
MonthlyIncome             -0.030160
NumCompaniesWorked         0.042831
PercentSalaryHike          0.033153
StandardHours                   NaN
StockOptionLevel          -0.008164
TotalWorkingYears         -0.169670
TrainingTimesLastYear     -0.047586
YearsAtCompany            -0.133003
YearsSinceLastPromotion   -0.031423
YearsWithCurrManager      -0.154692
Name: Attrition, dtype: float64

### Correlation and p-value for Attrition vs other variables using pearsonr()
Correlation analysis is used to find the interdependency of variables.

In [10]:
from scipy.stats import pearsonr

In [13]:
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EmployeeCount', 'EmployeeID', 'Gender',
       'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')

### Attrition and Age
- H0 = There is no significant correlation between Attrition and Age
- Ha = There is significant correlation between Attrition and Age 

In [16]:
r, p = pearsonr(df.Attrition, df.Age)
print(r, p)
if p < 0.05:
    print("H0 is rejected")
    print("Ha is accepted")
else:
    print("H0 is accepted")
    print("Ha is rejected")

-0.15839867954096706 5.1265982193780794e-26
H0 is rejected
Ha is accepted


### Inference 
- As we can see p < 0.05 Ha is accepted
- r = -0.15, there is low negative correlation between Attrition and Age

### Attrition and DistanceFromHome
- H0 = There is no significant correlation between Attrition and DistanceFromHome
- Ha = There is significant correlation between Attrition and DistanceFromHome

In [17]:
r, p = pearsonr(df.Attrition, df.DistanceFromHome)
print(r, p)
if p < 0.05:
    print("H0 is rejected")
    print("Ha is accepted")
else:
    print("H0 is accepted")
    print("Ha is rejected")

-0.009448638515156243 0.5317715668019558
H0 is accepted
Ha is rejected


### Inference 
- As we can see p >= 0.05 H0 is accepted
- r = -0.009, there is low negative correlation between Attrition and DistanceFromHome

### Attrition and MonthlyIncome
- H0 = There is no significant correlation between Attrition and MonthlyIncome
- Ha = There is significant correlation between Attrition and MonthlyIncome

In [18]:
r, p = pearsonr(df.Attrition, df.MonthlyIncome)
print(r, p)
if p < 0.05:
    print("H0 is rejected")
    print("Ha is accepted")
else:
    print("H0 is accepted")
    print("Ha is rejected")

-0.030160293808460664 0.045890862744719166
H0 is rejected
Ha is accepted


### Inference 
- As we can see p < 0.05 Ha is accepted
- r = -0.0301, there is low negative correlation between Attrition and MonthlyIncome

### Attrition and TotalWorkingYears
- H0 = There is no significant correlation between Attrition and TotalWorkingYears
- Ha = There is significant correlation between Attrition and TotalWorkingYears

In [19]:
r, p = pearsonr(df.Attrition, df.TotalWorkingYears)
print(r, p)
if p < 0.05:
    print("H0 is rejected")
    print("Ha is accepted")
else:
    print("H0 is accepted")
    print("Ha is rejected")

-0.1696699168472392 1.1645434967091854e-29
H0 is rejected
Ha is accepted


### Inference 
- As we can see p < 0.05 Ha is accepted
- r = -0.17, there is low negative correlation between Attrition and TotalWorkingYears

### Attrition and YearsAtCompany
- H0 = There is no significant correlation between Attrition and YearsAtCompany
- Ha = There is significant correlation between Attrition and YearsAtCompany

In [20]:
r, p = pearsonr(df.Attrition, df.YearsAtCompany)
print(r, p)
if p < 0.05:
    print("H0 is rejected")
    print("Ha is accepted")
else:
    print("H0 is accepted")
    print("Ha is rejected")

-0.13300261842521544 9.476118084836507e-19
H0 is rejected
Ha is accepted


### Inference 
- As we can see p < 0.05 Ha is accepted
- r = -0.133, there is low negative correlation between Attrition and YearsAtCompany

### Attrition and YearsWithCurrManager
- H0 = There is no significant correlation between Attrition and YearsWithCurrManager
- Ha = There is significant correlation between Attrition and YearsWithCurrManager

In [21]:
r, p = pearsonr(df.Attrition, df.YearsWithCurrManager)
print(r, p)
if p < 0.05:
    print("H0 is rejected")
    print("Ha is accepted")
else:
    print("H0 is accepted")
    print("Ha is rejected")

-0.15469153690287274 7.105369646771178e-25
H0 is rejected
Ha is accepted


### Inference 
- As we can see p < 0.05 Ha is accepted
- r = -0.155, there is low negative correlation between Attrition and YearsWithCurrManager